definitions/features/instance.rb

```diff
@@ -44,4 +49,101 @@ def postgresql_local?
   end
+
+  def ping?
+    if feature(:katello)
+      katello_ping
+    elsif external_proxy?
+      proxy_ping
+    else
+      foreman_ping
+    end
+    last_ping_result
+  end
+
+  def server_connection
+    net = Net::HTTP.new(ForemanMaintain.config.foreman_url, ForemanMaintain.config.foreman_port)
+    net.use_ssl = true
+    net
+  end
+
+  private
+
+  # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
+  def katello_ping
+    res = server_connection.get('/katello/api/ping')
+    logger.debug('Called /katello/api/ping')
+    logger.debug("Response: #{res.code}, #{res.body}")
+    response = JSON.parse(res.body)
+    if res.code != '200' # foreman error
+      set_ping_result(false, response.message, nil)
+    else # valid response
+      failing_components = pick_failing_components(response['services'])
+      if failing_components.empty? # all okay
+        set_ping_result(true, 'Success', nil)
+      else # some components not okay
+        set_ping_result(false,
+                        "Some components are failing: #{failing_components.join(', ')}",
+                        component_services(failing_components))
+      end
+    end
+  rescue StandardError => e # server error, server down
+    set_ping_result(false, "Couldn't connect to the server: #{e.message}", nil)
+  end
+  # rubocop:enable Metrics/MethodLength, Metrics/AbcSize
+
+  def foreman_ping
+    res = server_connection.get('/apidoc/apipie_checksum')
+    logger.debug('Called /apidoc/apipie_checksum')
+    logger.debug("Response: #{res.code}, #{res.body}")
+    if res.code != '200' # foreman error
+      set_ping_result(false, response.message, nil)
+    else # valid response
+      set_ping_result(true, 'Success', nil)
+    end
+  rescue StandardError => e # server error, server down
+    set_ping_result(false, "Couldn't connect to the server: #{e.message}", nil)
+  end
+
+  def proxy_ping
+    feature(:foreman_proxy).features
+    set_ping_result(true, 'Success', nil)
+  rescue StandardError => e # server error, proxy down
+    set_ping_result(false, "Couldn't connect to the proxy: #{e.message}", nil)
+  end
+
+  def pick_failing_components(components)
+    components.inject([]) do |failing, (name, data)|
+      data['status'] != 'ok' ? failing << name : failing
+    end
+  end
+
+  def set_ping_result(result, message, failing_services)
+    @last_ping_status = message
+    @last_ping_failing_services = failing_services
+    @last_ping_result = result
+  end
+
+  def installer_scenario_answers
+    feature(:installer).answers
+  end
+
+  def component_features_map
+    {
+      'candlepin_auth' =>  %w[candlepin candlepin_database],
+      'candlepin' => %w[candlepin candlepin_database],
+      'pulp_auth' => %w[pulp mongo],
+      'pulp' => %w[pulp mongo],
+      'foreman_tasks' => %w[foreman_tasks]
+    }
+  end
+
+  def component_services(components)
+    components = [components].flatten(1)
```

Comment:
If the intend is to make sure the `components` is always array, using `Array(components)` would feel a bit better to me

Reply:
Updated